auslib/db.py

```diff
@@ -665,9 +718,243 @@ def rollbackChange(self, change_id, changed_by, transaction=None):
+            try:
+                time.gmtime(conditions["when"])
+            except:
+                raise ValueError("Cannot parse 'when' as a unix timestamp.")
+
+    def insert(self, changed_by, transaction=None, dryrun=False, **columns):
+        base_table_where = []
+        for pk in self.base_primary_key:
+            base_column = getattr(self.baseTable, pk)
+            if pk in columns:
+                # If a non-null data_version was provided it implies that the
+                # base table row should already exist. This will be checked for
+                # after we finish basic checks on the individual parts of the
+                # PK.
+                if "data_version" in columns and columns["data_version"]:
+                    base_table_where.append(getattr(self.baseTable, pk) == columns[pk])
+            # Non-Integer columns can have autoincrement set to True for some reason.
+            # Any non-integer columns in the primary key are always required (because
+            # autoincrement actually isn't a thing for them), and any Integer columns
+            # that _aren't_ autoincrement are required as well.
+            elif not isinstance(base_column.type, (sqlalchemy.types.Integer,)) or not base_column.autoincrement:
+                raise ValueError("Missing primary key column '%s' which is not autoincrement", pk)
+
+        # If anything ended up in base_table_where, it means that the baseTable
+        # row should already exist. In these cases, we need to check to make sure
+        # that the scheduled change has the same data version as the base table,
+        # to ensure that a change is not being scheduled from an out of date version
+        # of the base table row.
+        if base_table_where:
+            current_data_version = self.baseTable.select(columns=(self.baseTable.data_version,), where=base_table_where, transaction=transaction)
+
+            if not current_data_version:
+                raise ValueError("Cannot create scheduled change with data_version for non-existent row")
+
+            if current_data_version and current_data_version[0]["data_version"] != columns.get("data_version"):
+                raise OutdatedDataError("Wrong data_version given for base table, cannot create scheduled change.")
+
+        what = self._prefixColumns(columns)
+        conditions = {}
+        for col in what:
+            if not col.startswith("base_"):
+                conditions[col] = what[col]
+
+        self._validateConditions(conditions)
+
+        # Use the appropriate base table methods in dry run mode to ensure the
+        # user has permission for the change they want to schedule
+        if columns.get("data_version"):
+            self.baseTable.update(base_table_where, columns, changed_by, columns["data_version"], transaction=transaction, dryrun=True)
+        else:
+            self.baseTable.insert(changed_by, transaction=transaction, dryrun=True, **columns)
```

Comment:
We could reduce some duplication here, but the only thing that would truly be factored out (that is to say: making a decision in one place instead of many) is the data version check to decide if its an update or insert. I could go either way on this one.